### PR TITLE
Upgrade to aruba 0.8.

### PR DESCRIPTION
**Pull request by myronmarston**

### CODE REVIEW

**Comment by maxmeyer** [comment]
Mmmh... Will have a look tomorrow.

File: spec/integration/filtering_spec.rb [hunk 1]
```diff
@@ -178,7 +178,7 @@ def run_rerun_command_for_failing_spec
       expect(last_cmd_stdout).to match(/3 examples, 0 failures/)
 
       # Using absolute paths...
-      spec_root = in_current_dir { File.expand_path("spec") }
+      spec_root = cd('.') { File.expand_path("spec") }
```
Comment (dg-ratiodata) [inline spec/integration/filtering_spec.rb]: You better should use `expand_path("spec")` here

**Comment by maxmeyer** [comment]
BTW: We're working on running tests with appveyor to have reliable builds on windows as well.

**Comment by myronmarston** [comment]
@maxmeyer / @dg-ratiodata -- now we're getting a new failure from aruba:

```
  1) Persistence failures when `config.example_status_persistence_file_path` is configured to a file path for which we lack permissions emits a helpful warning to the user, indicating we cannot read from it, and still runs the spec suite
     Failure/Error: write_file_formatted "spec/1_spec.rb", "
     NoMethodError:
       undefined method `write' for File:Class
     # /home/travis/build/rspec/bundle/ruby/1.9.1/gems/aruba-0.8.1/lib/aruba/creators/aruba_file_creator.rb:25:in `write'
     # /home/travis/build/rspec/bundle/ruby/1.9.1/gems/aruba-0.8.1/lib/aruba/api/filesystem.rb:123:in `write_file'
     # ./spec/support/aruba_support.rb:49:in `write_file_formatted'
     # ./spec/integration/persistence_failures_spec.rb:33:in `block (4 levels) in <top (required)>'
     # ./lib/rspec/core/example.rb:386:in `instance_exec'
     # ./lib/rspec/core/example.rb:386:in `instance_exec'
     # ./lib/rspec/core/hooks.rb:357:in `run'
     # ./lib/rspec/core/hooks.rb:521:in `block in run_owned_hooks_for'
     # ./lib/rspec/core/hooks.rb:520:in `each'
     # ./lib/rspec/core/hooks.rb:520:in `run_owned_hooks_for'
     # ./lib/rspec/core/hooks.rb:607:in `block in run_example_hooks_for'
     # ./lib/rspec/core/hooks.rb:606:in `reverse_each'
     # ./lib/rspec/core/hooks.rb:606:in `run_example_hooks_for'
     # ./lib/rspec/core/hooks.rb:476:in `run'
     # ./lib/rspec/core/example.rb:433:in `run_before_example'
     # ./lib/rspec/core/example.rb:211:in `block in run'
     # ./lib/rspec/core/example.rb:439:in `block in with_around_and_singleton_context_hooks'
     # ./lib/rspec/core/example.rb:396:in `block in with_around_example_hooks'
     # ./lib/rspec/core/hooks.rb:478:in `block in run'
     # ./lib/rspec/core/hooks.rb:618:in `block in run_around_example_hooks_for'
     # ./lib/rspec/core/example.rb:281:in `call'
     # ./lib/rspec/core/example.rb:281:in `call'
     # ./spec/support/sandboxing.rb:14:in `block (3 levels) in <top (required)>'
     # ./lib/rspec/core/sandbox.rb:29:in `sandboxed'
     # ./spec/support/sandboxing.rb:7:in `block (2 levels) in <top (required)>'
     # ./lib/rspec/core/example.rb:386:in `instance_exec'
     # ./lib/rspec/core/example.rb:386:in `instance_exec'
     # ./lib/rspec/core/hooks.rb:389:in `execute_with'
     # ./lib/rspec/core/hooks.rb:620:in `block (2 levels) in run_around_example_hooks_for'
     # ./lib/rspec/core/example.rb:281:in `call'
     # ./lib/rspec/core/example.rb:281:in `call'
     # ./lib/rspec/core/hooks.rb:619:in `run_around_example_hooks_for'
     # ./lib/rspec/core/hooks.rb:478:in `run'
     # ./lib/rspec/core/example.rb:396:in `with_around_example_hooks'
     # ./lib/rspec/core/example.rb:439:in `with_around_and_singleton_context_hooks'
     # ./lib/rspec/core/example.rb:209:in `run'
     # ./lib/rspec/core/example_group.rb:559:in `block in run_examples'
     # ./lib/rspec/core/example_group.rb:555:in `map'
     # ./lib/rspec/core/example_group.rb:555:in `run_examples'
     # ./lib/rspec/core/example_group.rb:521:in `run'
     # ./lib/rspec/core/example_group.rb:522:in `block in run'
     # ./lib/rspec/core/example_group.rb:522:in `map'
     # ./lib/rspec/core/example_group.rb:522:in `run'
     # ./lib/rspec/core/example_group.rb:522:in `block in run'
     # ./lib/rspec/core/example_group.rb:522:in `map'
     # ./lib/rspec/core/example_group.rb:522:in `run'
     # ./lib/rspec/core/runner.rb:115:in `block (3 levels) in run_specs'
     # ./lib/rspec/core/runner.rb:115:in `map'
     # ./lib/rspec/core/runner.rb:115:in `block (2 levels) in run_specs'
     # ./lib/rspec/core/configuration.rb:1632:in `with_suite_hooks'
     # ./lib/rspec/core/runner.rb:114:in `block in run_specs'
     # ./lib/rspec/core/reporter.rb:77:in `report'
     # ./lib/rspec/core/runner.rb:113:in `run_specs'
     # ./lib/rspec/core/runner.rb:89:in `run'
     # ./lib/rspec/core/runner.rb:73:in `run'
     # ./lib/rspec/core/runner.rb:41:in `invoke'
     # ./exe/rspec:4:in `<top (required)>'
     # script/rspec_with_simplecov:42:in `load'
     # script/rspec_with_simplecov:42:in `<main>'
```

It's failing on 1.9.2.  `File.write` was a new API added in MRI 1.9.3.  Does aruba support 1.9.2?  RSpec still does...

**Comment by maxmeyer** [comment]
Mmmh... I checked on 1.9 only. I added 1.9.2 to travis build and modified the check. Jesus... I didn't expect that pain on upgrading aruba....

**Comment by maxmeyer** [comment]
I pushed the change to aruba/master. Can you try this one?

**Comment by myronmarston** [comment]
@maxmeyer -- your changes to aruba master fixed 1.9.2, but now we're getting a ton of warnings on JRuby that we don't get from aruba 0.6.2.  Any ideas?

**Comment by dg-ratiodata** [comment]
One thing is a warning generated from a new gem we're using, I filed an issue for that https://github.com/egonSchiele/contracts.ruby/issues/178. But the HOME-warning, is new to me. I had quite strange issues with PATH on jruby I worked around. I'm not sure, there seems to be a problem with ENV.  Maybe it's even a thread-issue, but that's really a guess.

**Comment by dg-ratiodata** [comment]
I will try to get jruby running on my virtual machine... Maybe I can reproduce the error. Although I could not reproduce the PATH issue with jruby locally.

**Comment by dg-ratiodata** [comment]
"Perfect." At least I can reproduce the HOME-variable warning.

**Comment by dg-ratiodata** [comment]
I've some troubles to reproduce the failure when I run specific example. You can safely ignore the env vars. I only need them to make jruby run and do not "destroy" my mri-gems.

```
GEM_ROOT=~/.gem/jruby/ GEM_PATH=~/.gem/jruby/ GEM_HOME=~/.gem/jruby/ jruby --2.0 script/rspec_with_simplecov                    # Fails
GEM_ROOT=~/.gem/jruby/ GEM_PATH=~/.gem/jruby/ GEM_HOME=~/.gem/jruby/ jruby --2.0 script/rspec_with_simplecov ./spec/            # Fails
GEM_ROOT=~/.gem/jruby/ GEM_PATH=~/.gem/jruby/ GEM_HOME=~/.gem/jruby/ jruby --2.0 script/rspec_with_simplecov ./spec/rspec/      # Succeeds
GEM_ROOT=~/.gem/jruby/ GEM_PATH=~/.gem/jruby/ GEM_HOME=~/.gem/jruby/ jruby --2.0 script/rspec_with_simplecov ./spec/rspec/core/ # Succeeds
```

Using the new bisect functionality `rspec` told me, the errors do not depend on the order. Is the "true" or just a hint?

```
JRUBY_OPTS="-Xcext.enabled=true" GEM_ROOT=~/.gem/jruby/ GEM_PATH=~/.gem/jruby/ GEM_HOME=~/.gem/jruby/ jruby --2.0 script/rspec_with_simplecov --seed 123 --bisect
Bisect started using options: "--seed 123"
Running suite to find failures... (2 minutes 21.7 seconds)
Starting bisect with 15 failing examples and 1767 non-failing examples.
Checking that failure(s) are order-dependent... failure is not order-dependent

Bisect complete! Reduced necessary non-failing examples from 1767 to 0 in 14.14 seconds.

The minimal reproduction command is:
  rspec './spec/integration/filtering_spec.rb[1:1,1:2:1,1:3:1,1:4:1,1:4:2,1:5:1,1:5:2]' './spec/integration/order_spec.rb[1:1:1,1:2:1,1:3:1,1:3:2,1:4:1,1:5:1]' './spec/integration/persistence_failures_spec.rb[1:1:1:1,1:1:2:1]' --seed 123
```

**Comment by maxmeyer** [comment]
Ok. Our test suite now runs on Ruby 1.9.2 as well without any problems. I'm still working on the HOME-variable thing.

**Comment by maxmeyer** [comment]
Can it be that simple? 

```ruby
require 'rspec/expectations'

module RSpec
  class Scope
    include RSpec::Matchers
  end

  def self.scope(&block)
    Scope.new.instance_eval(&block)
  end
end
```

**Comment by dg-ratiodata** [comment]
I fixed the HOME-variable problem. It was a bug in `Jruby` - see jruby/jruby#3162 and https://github.com/reuben-sutton/jruby/commit/4fad5ded23de90a780b150253d8664cf6f43dc8e. I worked around that one and rspec and cucumber test suite runs fine on my local machine. Only `rubocop` complains.

**Comment by myronmarston** [comment]
What do I have to do to get the fixes?

**Comment by dg-ratiodata** [comment]
`#dup` on `ENV.to_h`. :smile:

**Comment by dg-ratiodata** [comment]
Damn.... Didn't understand your question first. Use aruba#master.

**Comment by myronmarston** [comment]
I kicked the build.  Now it's failing with:


```
(::) failed steps (::)
undefined method `unescape' for Aruba::Platforms::UnixPlatform:Class (NoMethodError)
./features/step_definitions/additional_cli_steps.rb:106:in `/^I have run `([^`]*)`$/'
features/command_line/init.feature:27:in `And I have run `rspec --init`'
Failing Scenarios:
cucumber features/command_line/init.feature:25 # Scenario: Accept and use the recommended settings in `spec_helper` (which are initially commented out)
```

**Comment by dg-ratiodata** [comment]
Oh. Sorry. We moved from `Aruba::Platform` to `Aruba.platform`. I didn't realize, that you configured aruba in your '.gemspec'. Sorry for the false positive information about the passing build. I didn't check my local Gemfile properly.

Now I'm running the build with:
~~~bundler
GIT
  remote: https://github.com/cucumber/aruba
  revision: 507b66327d21bbbb61dc1c97a28f79aa43a6e62f
  specs:
    aruba (0.8.2)
      childprocess (~> 0.5.6)
      contracts (~> 0.9)
      cucumber (>= 1.3.19)
      ffi (~> 1.9.10)
      rspec-expectations (>= 2.99)
~~~

The result is this:
~~~
[git:update-aruba]% be rake
/usr/bin/ruby -w -I/home/xgvndeg/.gem/ruby/2.2.2/bundler/gems/rspec-support-f82e762eb550/lib:/home/xgvndeg/data/projects/rspec-core/lib /home/xgvndeg/data/projects/rspec-core/exe/rspec --pattern spec/\*\*\{,/\*/\*\*\}/\*_spec.rb
Run options:
  include {:focus=>true}
  exclude {:ruby=>#<Proc:./spec/spec_helper.rb:102>}

All examples were filtered out; ignoring {:focus=>true}

Randomized with seed 9334
....................................................................................................................................*......................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................*...................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................*..........................................................................................................

Pending: (Failures listed here are expected and do not affect your suite's status)

  1) RSpec::Core::Formatters::Helpers format duration with mathn loaded returns 'x minutes xx.x seconds' formatted string
     # lib/mathn.rb is deprecated in Ruby 2.2
     # ./spec/rspec/core/formatters/helpers_spec.rb:58

  2) RSpec::Core::Metadata for an example points :example_group to the same hash object as other examples in the same group
     # Cannot maintain this and provide full `:example_group` backwards compatibility (see GH #1490):(
     Failure/Error: expect(b[:description]).to eq("new description")

       expected: "new description"
            got: "group"

       (compared using ==)
     # ./spec/rspec/core/metadata_spec.rb:138:in `block (3 levels) in <module:Core>'
     # ./spec/support/sandboxing.rb:14:in `block (3 levels) in <top (required)>'
     # ./spec/support/sandboxing.rb:7:in `block (2 levels) in <top (required)>'

  3) RSpec::Core::Formatters::HtmlFormatter produced HTML with mathn loaded is identical to the one we designed manually
     # lib/mathn.rb is deprecated in Ruby 2.2
     # ./spec/rspec/core/formatters/html_formatter_spec.rb:105


Finished in 15.96 seconds (files took 0.83086 seconds to load)
1802 examples, 0 failures, 3 pending

Randomized with seed 9334

/usr/bin/ruby -S bundle exec cucumber
Using the default profile...
.................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................

270 scenarios (270 passed)
1318 steps (1318 passed)
3m2.277s
~~~

You need to change this:

~~~diff
 Given /^I have run `([^`]*)`$/ do |cmd|
   fail_on_error = true
-  run_simple(Aruba::Platform.unescape(cmd), fail_on_error)
+  run_simple(Aruba.platform.unescape(cmd), fail_on_error)
 end
~~~

**Comment by maxmeyer** [comment]
I think `#unescape` has some value and I'm going to move it to `#extract_text`. And just a warning: Every method from `platform` should not be considered to be part of the public api of aruba: https://github.com/cucumber/aruba/blob/master/lib/aruba/platforms/unix_platform.rb#L10. I think we need to be more strict about that. For now you can leave the method above.

**Comment by maxmeyer** [comment]
I will also have a look into your build failures on JRUBY tomorrow.

**Comment by myronmarston** [comment]
> I think #unescape has some value and I'm going to move it to #extract_text. And just a warning: Every method from platform should not be considered to be part of the public api of aruba: https://github.com/cucumber/aruba/blob/master/lib/aruba/platforms/unix_platform.rb#L10. I think we need to be more strict about that. For now you can leave the method above.

We're only using `Aruba::Platform` because @dg-ratiodata introduced it in 2a972dc653ade00b8cc935b166653708f4f5b530 -- and now you're telling us it's not public and we can't rely on it?  In master we are using `unescape`, which is a clearly documented public API of aruba 0.6.2:

https://github.com/cucumber/aruba/blob/v0.6.2/lib/aruba/api.rb#L459-L470

...and in exchange for spending hours trying to upgrade, we are now apparently migrating to a less stable API.

This upgrade process for Aruba is taking more of my time than I've ever spent dealing with Aruba in the 4-5 years I've worked on RSpec -- and aruba 0.6.2 is working just fine for us.  And the benefit of upgrading seems to be an unstable API.  At this point, I can no longer devote time my personal time to upgrading us, and I don't think I want to upgrade while aruba is in this period of API instability.  I'll be happy to try again when you've made some sort of public commitment to API stability (such as SemVer -- presumably in 1.0?) but until then we can just sit tight on an old version.  If one of you wants to put the effort into upgrading us and keeping us current against new Aruba versions I'd be happy to merge those PRs but I don't have the time to do that myself.

Closing.

**Comment by maxmeyer** [comment]
Sorry, @myronmarston that it is such a unpleasant upgrade procedure. I really did not foresee that. 

And to the `Aruba::Platform`-thing. @dg-ratiodata and @maxmeyer are the same person - me - I need to use a separate account during normal business hours. Sorry for giving such a  recommendation and submit private API-code to `rspec`.

I am going to fork `rspec-core` and activate `travis` to get some better feedback for different rubies. `JRuby` really gave me quite a hard time. Unfortunately I'm not able to reproduce all issues on my local machine. Maybe there's even another bug in `child_process` on `jruby`. :-/

**Pull request closed without merging**